pkg/insightstore/chunk.go

```diff
@@ -0,0 +1,284 @@
+}
+
+func (c ChangeFailureRate) Value() float32 {
+	return c.Rate
+}
+
+type Chunk interface {
+	// GetFilePath gets filepath
+	GetFilePath() string
+	// SetFilePath sets filepath
+	SetFilePath(path string)
+	// GetAccumulatedTo gets AccumulatedTo
+	GetAccumulatedTo() int64
+	// SetAccumulatedTo sets AccumulatedTo
+	SetAccumulatedTo(a int64)
+	// GetDataPoint gets list of data points of specify step
+	GetDataPoints(step model.InsightStep) ([]DataPoint, error)
+	// DataCount returns count of data in specify step
+	DataCount(step model.InsightStep) int
+}
+
+// convert types to Chunk.
+func toChunk(i interface{}) (Chunk, error) {
+	switch p := i.(type) {
+	case *DeployFrequencyChunk:
+		return p, nil
+	case *ChangeFailureRateChunk:
+		return p, nil
+	default:
+		return nil, fmt.Errorf("cannot convert to Chunk: %v", p)
+	}
+}
+
+type DataPoint interface {
+	// Value gets data for model.InsightDataPoint
+	Value() float32
+	// Timestamp gets timestamp
+	GetTimestamp() int64
+}
+
+// convert types to list of DataPoint.
+func toDataPoints(i interface{}) ([]DataPoint, error) {
+	switch dps := i.(type) {
+	case []DeployFrequency:
+		dataPoints := make([]DataPoint, len(dps))
+		for j, dp := range dps {
+			dataPoints[j] = dp
+		}
+		return dataPoints, nil
+	case []ChangeFailureRate:
+		dataPoints := make([]DataPoint, len(dps))
+		for j, dp := range dps {
+			dataPoints[j] = dp
+		}
+		return dataPoints, nil
+	default:
+		return nil, fmt.Errorf("cannot convert to DataPoints: %v", dps)
+	}
+}
+
+type Chunks []Chunk
+
+func (cs Chunks) ExtractDataPoints(step model.InsightStep, from time.Time, count int) ([]*model.InsightDataPoint, error) {
+	var out []*model.InsightDataPoint
+	var to time.Time
+	switch step {
+	case model.InsightStep_YEARLY:
+		to = from.AddDate(count-1, 0, 0)
+	case model.InsightStep_MONTHLY:
+		to = from.AddDate(0, count-1, 0)
+	case model.InsightStep_WEEKLY:
+		to = from.AddDate(0, 0, (count-1)*7)
+	case model.InsightStep_DAILY:
+		to = from.AddDate(0, 0, count-1)
+	}
+
+	for _, c := range cs {
+		idp, err := extractDataPoints(c, step, from, to)
+		if err != nil {
+			return nil, err
+		}
+
+		out = append(out, idp...)
+
+		count = count - c.DataCount(step)
+
+		nextMonth := time.Date(from.Year(), from.Month()+1, 1, 0, 0, 0, 0, time.UTC)
+		from = normalizeTime(nextMonth, step)
+		if step == model.InsightStep_WEEKLY && from.Month() != nextMonth.Month() {
+			from = from.AddDate(0, 0, 7)
+		}
+	}
+
+	return out, nil
+}
+
+func extractDataPoints(chunk Chunk, step model.InsightStep, from, to time.Time) ([]*model.InsightDataPoint, error) {
+	target, err := chunk.GetDataPoints(step)
+	if err != nil {
+		return []*model.InsightDataPoint{}, nil
+	}
+
+	var result []*model.InsightDataPoint
+	for _, d := range target {
+		ts := d.GetTimestamp()
+		if ts <= to.Unix() && ts >= from.Unix() {
+			result = append(result, &model.InsightDataPoint{
+				Timestamp: ts,
+				Value:     d.Value(),
+			})
+		}
+		if ts > to.Unix() {
+			break
+		}
```

Comment:
nit:

``` golang
if ts > to.Unix() || ts < from.Unix() {
   break
}

result = append(result, &model.InsightDataPoint{
})
```

Reply:
List is sorted, but, `from` does not always specify the oldest data in the list. 

So, I think it's wrong to break when `ts < from.Unix()`, and we should have `if ts >= from.Unix()` to exclude data that older than `from`.

For the above reasons, I finally made the following changes

```
	for _, d := range target {
		ts := d.GetTimestamp()
		if ts > to.Unix() {
			break
		}

		if ts >= from.Unix() {
			result = append(result, &model.InsightDataPoint{
				Timestamp: ts,
				Value:     d.Value(),
			})
		}
	}
```